File: Scala/main/appendix/ThreadSafe.scala
package appendix

object ThreadSafe:
   class SafeSet[A]:
      private var elements = Set.empty[A]

      def += (elem: A): Int = synchronized {
         elements += elem
         elements.size
      }

      def all: Set[A] = synchronized(elements)
   end SafeSet
end ThreadSafe

File: Kotlin/main/appendix/ThreadSafe.kt
package appendix

object ThreadSafe {
  object SafeSet1 {
    class SafeSet<A> {
      private val elements = mutableSetOf<A>()

      @Synchronized
      fun add(elem: A): Int {
        elements += elem
        return elements.size
      }

      @Synchronized
      fun all(): Set<A> = elements.toSet() // full copy here
    }
  }

  object SafeSet2 {
    class SafeSet<A> {
      private var elements = setOf<A>()

      @Suppress("SuspiciousCollectionReassignment")
      @Synchronized
      fun add(elem: A): Int {
        elements += elem // full copy here
        return elements.size
      }

      @Synchronized
      fun all(): Set<A> = elements
    }
  }
}
